### ISSUE TITLE
Compiler freezes and never ends when if optional type is used as generic argument

### ISSUE THREAD

**Issue by prantlf** [vlang/v#18331]
### Describe the bug

When compiling the following code:

```go
fn f[T]() T {
	return none
}

r := f[?int]()
println(r)
```

The compiler will be compiling (or doing something else ;-) forever.

The code above was obtained when stripping the following code:

```go
pub fn unmarshal_text[T](input string, opts MarshalOpts) !T { ... }

fn test_unmarshal_empty_optional_input() {
	yaml.unmarshal_text[?int]('', yaml.MarshalOpts{}) or {
		assert false
	}
}
```

### Expected Behavior

The compiler should finish sometime :-) I expect the function to be resolved to this:

```go
fn f[?int]() ?int {
	return none
}
```

### Current Behavior

The compiler will never finish and no output will be written on the console.

### Reproduction Steps

1. Save the code above to a file `bug_optional.v`.
2. Execute `v bug_optional.v`.

I didn't dare to create an example at vosca.dev. Your server might start hating me :-)

### Possible Solution

I wish I knew :-)

### Additional Information/Context

There might be some endless recursion, when the optional type is passed to the generics resolver. If I use a concrete type as the generic argument and introduce the `?` first with the return type, the code will compile and run well:

```go
fn f[T]() ?T {
	return none
}

r := f[int]()
println(r)
```

### V version

V 0.3.4 0fc33c6

### Environment details (OS name and version, etc.)

V full version: V 0.3.4 1be798b.0fc33c6
OS: macos, macOS, 13.3.1, 22E772610a
Processor: 12 cpus, 64bit, little endian, Intel(R) Core(TM) i7-8850H CPU @ 2.60GHz

getwd: /Users/prantlf/Sources/github/v-yaml
vexe: /usr/local/Cellar/vlang/0.3.4/libexec/v
vexe mtime: 2023-06-02 11:18:01

vroot: OK, value: /usr/local/Cellar/vlang/0.3.4/libexec
VMODULES: OK, value: /Users/prantlf/.vmodules
VTMP: OK, value: /tmp/v_501

Git version: git version 2.41.0
Git vroot status: weekly.2023.21-58-g0fc33c6f
.git/config present: true

CC version: Apple clang version 14.0.3 (clang-1403.0.22.14.1)
thirdparty/tcc status: thirdparty-macos-amd64 46662e20

**Comment by felipensp** [comment]
```V
fn f[T]() ?T {
	return none
}

r := f[int]()
println(r)
```
works

**Comment by JalonSolov** [comment]
I just tried your example on Linux, and it immediately printed output and exited.
```
[jalon@7950x ~]$ v run foo.v
Option(none)
[jalon@7950x ~]$ v version
V 0.3.4 0fc33c6
[jalon@7950x ~]$ 
```

**Comment by prantlf** [comment]
Oh, thank God! And I mean nothing bad :-)

**Comment by felipensp** [comment]
I think it makes no sense to specify the concrete type as option, since it must be specified in the return type as `?T` to make clear that the function returns Option. So, I've submitted a PR which makes a parser erro such usage.

**Comment by prantlf** [comment]
Yes, I agree, @felipensp, that this wouldn't be needed in a reality. This type got resolved when I wrote a deserializer using generics and I used recursive functions. I changed the implementation to use runtime and compile-time conditions. Disallowing the syntax is fine for me.

I'll continue with these rules:

1) Calls like `deserialize[?int](...)` won't work. Probably nobody would pas an option type as a first-level deserialized type.

2) Usage of option type in arrays like `[]?int` should work. Somebody might want it, as a return type or a function returning multiple optional results. I cannot test it yet because of #18317.

3) Usage of option types in structs like `struct { answer ?int }` will work.
